src/transport/datamodel/utransport.rs

```diff
@@ -11,19 +11,61 @@
+#[async_trait]
 pub trait UTransport {
-    /// API to register the calling uE with the underlying transport implementation.
-    fn register(&self, uentity: UEntity, token: &[u8]) -> UStatus;
+    /// Authenticates with the underlying transport layer that the `uEntity` passed
+    /// matches the transport specific identity. This method requires a resolved `UUri`.
+    ///
+    /// # Arguments
+    /// * `uEntity` - Resolved `UEntity`.
+    ///
+    /// # Returns
+    /// Returns `OKSTATUS` if authentication was successful, `FAILSTATUS` if the calling `uE`
+    /// is not authenticated.
+    async fn authenticate(&self, entity: UEntity) -> UStatus;
 
-    /// Transmit UPayload to the topic using the attributes defined in UTransportAttributes.
-    fn send(&self, topic: UUri, payload: UPayload, attributes: UAttributes) -> UStatus;
+    /// Transmits `UPayload` to the topic using the attributes defined in `UTransportAttributes`.
+    ///
+    /// # Arguments
+    /// * `topic` - Resolved `UUri` topic to send the payload to.
+    /// * `payload` - Actual payload.
+    /// * `attributes` - Additional transport attributes.
+    ///
+    /// # Returns
+    /// Returns `OKSTATUS` if the payload has been successfully sent (ACK'ed), otherwise returns
+    /// `FAILSTATUS` with the appropriate failure.
+    async fn send(&self, topic: UUri, payload: UPayload, attributes: UAttributes) -> UStatus;
 
-    /// Register a method that will be called when a message comes in on the specific topic.
-    fn register_listener(&self, topic: UUri, listener: dyn UListener) -> UStatus;
+    /// Registers a listener to be called when `UPayload` is received for the specific topic.
+    ///
+    /// # Arguments
+    /// * `topic` - Resolved `UUri` for where the message arrived via the underlying transport technology.
+    /// * `listener` - The method to execute to process the data for the topic.
+    ///
+    /// # Returns
+    /// Returns `OKSTATUS` if the listener is registered correctly, otherwise returns `FAILSTATUS`
+    /// with the appropriate failure.
+    async fn register_listener(&self, topic: UUri, listener: Sender<UMessage>) -> UStatus;
 
-    /// Unregister a method on a topic. Messages arriving on this topic will no longer be processed by this listener.
-    fn unregister_listener(&self, topic: UUri, listener: dyn UListener) -> UStatus;
+    /// Unregister a listener for a given topic. Messages arriving on this topic will no longer be processed
+    /// by this listener.
+    ///
+    /// # Arguments
+    /// * `topic` - Resolved `UUri` for where the listener was registered to receive messages from.
+    /// * `listener` - The method to execute to process the data for the topic.
+    ///
+    /// # Returns
+    /// Returns `OKSTATUS` if the listener is unregistered correctly, otherwise returns `FAILSTATUS`
+    /// with the appropriate failure.
+    async fn unregister_listener(&self, topic: UUri, listener: Sender<UMessage>) -> UStatus;
```

Comment:
it doesn't have to be async

Reply:
I would leave it as `async` in the same way as `register_listener` is. 
What I would be careful about is that `Sender<UMessage>` comes from `std::sync::mpsc::Sender`, which is `sync`. That means that the `send` operation could block the thread. Therefore, I believe it's wiser to use an `async sender`. E.g.: [flume](https://docs.rs/flume/latest/flume/) provides both a `sync` and `async` channel.